File: src/utils/shiftLogic.js
/**
 * Pure functions to calculate and apply shifts on a 2D mesh grid.
 */

/**
 * Generates a perfectly square grid of size sqrt(p) x sqrt(p).
 * @param {number} p - Total number of nodes (must be perfect square)
 * @returns {Array<Array<Object>>} The initialized 2D grid.
 */
export function generateGrid(p) {
  const n = Math.sqrt(p);
  const grid = [];
  
  for (let r = 0; r < n; r++) {
    const row = [];
    for (let c = 0; c < n; c++) {
      const nodeId = r * n + c;
      row.push({
        id: nodeId,
        data: nodeId // Data initially matches the node's ID
      });
    }
    grid.push(row);
  }
  
  return grid;
}

/**
 * Shifts the data in each row by 'shift' amount.
 * Node i sends data to (i + shift) mod sqrt(p).
 * @param {Array<Array<Object>>} grid - Current 2D grid
 * @param {number} shift - The q_row shift amount
 * @returns {Array<Array<Object>>} A new 2D grid after row shifting
 */
export function rowShift(grid, shift) {
  const n = grid.length;
  const newGrid = [];
  
  for (let r = 0; r < n; r++) {
    const newRow = [];
    for (let c = 0; c < n; c++) {
      // The math trick slightly differs to avoid negative modulo in JS
      const srcC = ((c - shift) % n + n) % n;
      newRow.push({
        id: grid[r][c].id,
        data: grid[r][srcC].data
      });
    }
    newGrid.push(newRow);
  }
  
  return newGrid;
}

/**
 * Shifts the data in each column by 'shift' amount.
 * Node (r,c) sends data to (r + shift mod n, c).
 * @param {Array<Array<Object>>} grid - Current 2D grid
 * @param {number} shift - The q_col shift amount
 * @returns {Array<Array<Object>>} A new 2D grid after column shifting
 */
export function columnShift(grid, shift) {
  const n = grid.length;
  const newGrid = [];
  
  for (let r = 0; r < n; r++) {
    const newRow = [];
    for (let c = 0; c < n; c++) {
      const srcR = ((r - shift) % n + n) % n;
      newRow.push({
        id: grid[r][c].id,
        data: grid[srcR][c].data
      });
    }
    newGrid.push(newRow);
  }
  
  return newGrid;
}

/**
 * Performs and returns all stages of the 2-stage shift algorithm.
 * @param {number} p - Total nodes
 * @param {number} q - Total shift amount
 * @returns {Object} Object containing intermediate and final states
 */
export function meshShift(p, q) {
  const n = Math.sqrt(p);
  const qRow = q % n;
  const qCol = Math.floor(q / n);
  
  const initialGrid = generateGrid(p);
  const stage1Grid = rowShift(initialGrid, qRow);
  const stage2Grid = columnShift(stage1Grid, qCol);
  
  return {
    qRow,
    qCol,
    initialGrid,
    stage1Grid,
    stage2Grid
  };
}

File: python/shift_logic.py
import math

def generate_grid(p):
    """Generates a perfectly square grid of size sqrt(p) x sqrt(p)."""
    n = int(math.sqrt(p))
    grid = []
    for i in range(n):
        row = []
        for j in range(n):
            node_id = i * n + j
            # Each node stores its own un-shifted node_id as data initially
            row.append({'id': node_id, 'data': node_id})
        grid.append(row)
    return grid

def row_shift(grid, shift):
    """
    Shifts the data in each row by 'shift' amount.
    Node i sends data to i + shift mod sqrt(p).
    This means the data received at pos j comes from pos (j - shift) mod sqrt(p).
    """
    n = len(grid)
    new_grid = [[None for _ in range(n)] for _ in range(n)]
    
    for r in range(n):
        for c in range(n):
            src_c = (c - shift) % n
            new_grid[r][c] = {
                'id': grid[r][c]['id'],
                'data': grid[r][src_c]['data']
            }
            
    return new_grid

def column_shift(grid, shift):
    """
    Shifts the data in each column by 'shift' amount.
    Node (r, c) sends data to (r + shift mod n, c).
    """
    n = len(grid)
    new_grid = [[None for _ in range(n)] for _ in range(n)]
    
    for r in range(n):
        for c in range(n):
            src_r = (r - shift) % n
            new_grid[r][c] = {
                'id': grid[r][c]['id'],
                'data': grid[src_r][c]['data']
            }
            
    return new_grid

def mesh_shift(p, q):
    """
    Performs the full 2-stage shift algorithm.
    Stage 1: Row shift by (q mod sqrt(p))
    Stage 2: Column shift by floor(q / sqrt(p))
    """
    n = int(math.sqrt(p))
    
    q_row = q % n
    q_col = q // n
    
    print(f"Mesh size: {n}x{n} (p={p})")
    print(f"Target shift: {q}")
    print(f"Stage 1 (Row Shift): {q_row}")
    print(f"Stage 2 (Col Shift): {q_col}\n")
    
    grid = generate_grid(p)
    print("Initial Grid:")
    for row in grid:
        print([node['data'] for node in row])
        
    grid_stage1 = row_shift(grid, q_row)
    print("\nAfter Stage 1 (Row Shift):")
    for row in grid_stage1:
        print([node['data'] for node in row])
        
    grid_stage2 = column_shift(grid_stage1, q_col)
    print("\nAfter Stage 2 (Column Shift):")
    for row in grid_stage2:
        print([node['data'] for node in row])
        
    # Verify final positions
    correct = True
    for r in range(n):
        for c in range(n):
            node_id = r * n + c
            expected_data = (node_id - q) % p
            if grid_stage2[r][c]['data'] != expected_data:
                correct = False
                print(f"Mismatch at node {node_id}: expected {expected_data}, got {grid_stage2[r][c]['data']}")
                
    if correct:
        print("\nVerification Passed! All nodes have the correct data.")
    else:
        print("\nVerification Failed.")
        
    return grid_stage2

if __name__ == "__main__":
    # Test case: p=16, q=5
    # (Row shift = 5 mod 4 = 1)
    # (Col shift = 5 // 4 = 1)
    mesh_shift(16, 5)
